### ISSUE TITLE
Add RandomGaussianBlur

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/vision#2635]
## 🚀 Feature

Idea is to add a random gaussian blur image transform like in [SwAV](https://github.com/facebookresearch/swav/blob/d4970c83791c8bd9928ec4efcc25d4fd788c405f/src/multicropdataset.py#L65)

cc @vfdev-5

**Comment by tejank10** [comment]
Hi @vfdev-5, first-time contributor here. Can I take this up, if it is okay with you?

I see that `kornia` already has this transform [(link)](https://kornia.readthedocs.io/en/latest/_modules/kornia/filters/gaussian.html#gaussian_blur2d). Is it recommended to design this transform by using `kornia`'s utility or better to have those utilities natively in `torchvision`?

**Comment by vfdev-5** [comment]
Hi @tejank10 thanks for asking and proposing to work on this issue.

> Is it recommended to design this transform by using kornia's utility or better to have those utilities natively in torchvision?

In general it is better to avoid adding new dependencies (like `kornia`) for a single feature. This feature should be implemented for PIL and torch.tensor inputs. 

> first-time contributor here. Can I take this up, if it is okay with you?

Technically, this feature can be a bit complex for a first-time contributors but I can help to advance and iterate on sent PR etc.

There will be two implementations, see as example [`F.rgb_to_grayscale`](https://github.com/pytorch/vision/blob/5f616a2b962cac01b6fcb68096d752a2d8d4050e/torchvision/transforms/functional.py#L981). 
For torch.tensor input we have to use pytorch `F.conv2d` with gaussian kernel and for PIL we can use [GaussianBlur](https://pillow.readthedocs.io/en/stable/reference/ImageFilter.html?highlight=Gaussian#PIL.ImageFilter.GaussianBlur).

What do you think ?

**Comment by tejank10** [comment]
Thanks for the direction @vfdev-5.

I was comparing PIL's GaussianBlur functionality with the ones from other libraries. This will help in designing the transform for torch.Tensor and mitigate the possible inconsistencies between PIL vs tensor outputs.

PIL's [GaussianBlur](https://pillow.readthedocs.io/en/stable/reference/ImageFilter.html?highlight=Gaussian#PIL.ImageFilter.GaussianBlur) performs BoxFilter thrice to approximate GaussianBlur. It takes `sigma` as input. If we compare it with [OpenCV](https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#gac05a120c1ae92a6060dd0db190a61afa)'s GaussianBlur, it are more controllable. We can control the kernel size and padding mode as well there. It allows not to specify sigma and falls back to a value that is computed using kernel size.
Even in terms of the output, there are differences if we compare PIL's GaussianBlur with that of OpenCV.

While designing the blur for torch.Tensor should we go about it the OpenCV way, or try to match PIL since we plan to use PIL's version when PIL Image as input?

**Comment by vfdev-5** [comment]
@tejank10 for the first version I think PIL's GaussianBlur with a single parameter `radius` is OK. We can not use opencv as dependency neither.

> While designing the blur for torch.Tensor should we go about it the OpenCV way, or try to match PIL since we plan to use PIL's version when PIL Image as input?

Please, take a look how implemented [`F.rgb_to_grayscale`](https://github.com/pytorch/vision/blob/5f616a2b962cac01b6fcb68096d752a2d8d4050e/torchvision/transforms/functional.py#L981). Idea is to be able to dispatch according to the input type: if input is PIL image => `F_pil.gaussian_blur`, if input is torch tensor => `F_t.gaussian_blur`. `F_pil.gaussian_blur` should perform PIL's GaussianBlur and `F_t.gaussian_blur` should work directly on tensor without using any other library: a) create gaussian kernel tensor as it is done in PIL code and apply pytorch's `F.cond2d`...

**Comment by tejank10** [comment]
@vfdev-5 Yeah I understand that using any other library is not the way. :) 
I was trying to contrast the way Gaussian blur actually happens between PIL and OpenCV, and should our Gaussian kernel creation function be as customizable as that of OpenCV, or are we okay with that tradeoff and let the function perform the same as PIL.

**Comment by vfdev-5** [comment]
@tejank10 yes we are OK with that tradeoff and let the function perform the same as PIL.

**Comment by yaox12** [comment]
Glad to see someone is working on this! It would be better for me if the function is to be as controllable as that of OpenCV. 
I'm currently doing some research work on image self-supervised learning (SSL), where people usually perform Gaussian blur with fixed `kernel_size` while random `sigma`, for example, [SimCLR](https://github.com/sthalles/SimCLR/blob/master/data_aug/gaussian_blur.py) and [BYOL](https://github.com/yaox12/BYOL-PyTorch/blob/master/data/byol_transform.py#L17). 
So it would be more helpful if we can control both the `kernel_size` and `sigma`, instead of `radius` only. Here is a typical usage of Gaussian blur in SSL:

```python
class GaussianBlur():
    def __init__(self, kernel_size, sigma_min=0.1, sigma_max=2.0):
        self.sigma_min = sigma_min
        self.sigma_max = sigma_max
        self.kernel_size = kernel_size

    def __call__(self, img):
        sigma = np.random.uniform(self.sigma_min, self.sigma_max)
        img = cv2.GaussianBlur(np.array(img), (self.kernel_size, self.kernel_size), sigma)
        return Image.fromarray(img.astype(np.uint8))

transform = transforms.Compose([
    ...,
    transforms.RandomApply([GaussianBlur(kernel_size=23)], p=0.1),
    ...,
])
```
In summary, I hope we can provide:
1. function `F.gaussian_blur` takes `kernel_size` and `sigma` as inputs, just like `cv2.GaussianBlur`
2. class `transform.GaussianBlur` that accepts range inputs just like [albumentations](https://github.com/albumentations-team/albumentations/blob/master/albumentations/augmentations/transforms.py#L2504) and [`transforms.ColorJitter`](https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py#L995)

**Comment by vfdev-5** [comment]
@yaox12 thanks for the suggestion, let's see what could we do.

**Comment by tejank10** [comment]
[This](https://github.com/tejank10/vision/commit/e9ece08d666eaf608960f7bf185c7fafe57d41ea) is how it currently looks. It randomly chooses a kernel radius from the given range, creates kernel accordingly, and applies the blurring function.

Following is the error image between PIL vs Tensor versions when I tried it on the hopper image:
![erimg](https://user-images.githubusercontent.com/19868340/92108185-92e54400-ee04-11ea-844c-697e5ea07158.jpg)

Let me know your thoughts, @vfdev-5 , @yaox12 on what could be possible improvements.

**Comment by vfdev-5** [comment]
@tejank10 sorry for delayed reply. Could you please draft a PR based on your branch and we can iterate over the code.

**Comment by vfdev-5** [comment]
Closed via #2658
